src/datasets/utils/remote_utils.py

```diff
@@ -0,0 +1,177 @@
+
+        Returns:
+            requests.Response
+        """
+        headers = copy.deepcopy(headers) or {}
+        headers["user-agent"] = cls.get_datasets_user_agent(user_agent=headers.get("user-agent"))
+        response = cls.request_with_retry(
+            verb="HEAD",
+            url=url,
+            proxies=proxies,
+            headers=headers,
+            cookies=cookies,
+            allow_redirects=allow_redirects,
+            timeout=timeout,
+            max_retries=max_retries,
+        )
+        return response
+
+    @classmethod
+    def get(cls, url, proxies=None, resume_size=0, headers=None, cookies=None, max_retries=0, callback=None):
+        """Wrapper around `requests` GET.
+
+        Args:
+            url (str): The URL of the resource to fetch.
+            proxies (dict): Parameter passed to `requests.request` with the protocol to the URL of the proxy.
+            resume_size (int): Part of the resource to be returned, passed to Range HTTP request Header as bytes range
+                start.
+            headers (dict, optional): Parameter passed to `requests.request` with the HTTP Headers to be sent with the
+                Request.
+            cookies (dict, optional): Parameter passed to `requests.request` with the cookies to be sent with the
+                Request.
+            max_retries (int, default=0): Maximum number of retries, defaults to 0 (no retries).
+            callback (callable, optional): Callback function to be called with each response data chunk.
+
+        Returns:
+            generator or type returned by `callback`
+        """
+        headers = copy.deepcopy(headers) or {}
+        headers["user-agent"] = cls.get_datasets_user_agent(user_agent=headers.get("user-agent"))
+        if resume_size > 0:
+            headers["Range"] = "bytes=%d-" % (resume_size,)
+        response = cls.request_with_retry(
+            verb="GET",
+            url=url,
+            stream=True,
+            proxies=proxies,
+            headers=headers,
+            cookies=cookies,
+            max_retries=max_retries,
+        )
+        if response.status_code == 416:  # Range not satisfiable
+            return
+        if callback:
+            content_length = response.headers.get("Content-Length")
+            total = resume_size + int(content_length) if content_length is not None else None
+            not_verbose = bool(logger.getEffectiveLevel() > WARNING)
+            progress = tqdm(
+                unit="B",
+                unit_scale=True,
+                total=total,
+                initial=resume_size,
+                desc="Downloading",
+                disable=not_verbose,
+            )
+            for chunk in response.iter_content(chunk_size=1024):
+                if chunk:  # filter out keep-alive new chunks
+                    progress.update(len(chunk))
+                    callback(chunk)
```

Comment:
Why use callbacks here ? It used to write to the file directly right ?

I can understand that writing to a file is out of scope for a method named "get" but in this case `get_from_cache` can just take the iterator returned by `get` and take care of iterating through the chunks, update the progress bar and write ?

This way the one who's writing files is only `get_from_cache`, not HttpClient nor FtpClient

Reply:
Yes, you are right. I first implemented callback to keep the parallelism: after the the call, data was written. Afterwards I thought something similar to what you mention, once I realized that with callback=None I was returning a generator and iterating through it could be done outside!

I factorize this! ;)

Reply:
The same as below applies here: I must keep the connection open while reading the data. I am going to make a small refactoring to make this clear.